File: day15-1.pas
program day15p1(output);
{$I BIGILIB.PAS}
type
  str16 = string[16];
var
  hashfile: text;
  hashme: str16;
  hashval: integer;
  hashvalb, total: BigInt;

{ Compute hash value from string }
function hash(str:str16): integer;
var i,current: integer;
begin
  current := 0;
  for i := 1 to length(str) do
  begin
    current := current + ord(str[i]);
    current := current*17;
    current := current mod 256
  end;
  hash := current  { Function return }
end;

{ Read csv until end of line hit, retrun string }
function getstr: str16;
var hm: str16;
    ch: char;
begin
  hm := '';
  read(hashfile,ch);
  while (ch <> ',') and (not eoln(hashfile)) do
  begin
    hm := hm+ch;
    read(hashfile,ch)
  end;

  if eoln(hashfile) and (length(hm) > 0) then  { Deal with eoln hit }
    hm := hm+ch;
  getstr := hm  { Function return }
end;

begin
  writeln('Advent of Code 2023 Day 15, part 1');
  writeln('');

  assign(hashfile,'day15in.txt');
  reset(hashfile);
  total := '0';
  while not eof(hashfile) do
  begin
    hashme := getstr;
    if (hashme <> '') then
    begin
      hashval := hash(hashme);
      str(hashval,hashvalb);
      total := add(total,hashvalb)
    end
  end;
  close(hashfile);
  { Remove last has value as eoln hit }
  total := sub(total,hashvalb);
  writeln ('');
  writeln ('Sum of hash values is ',total)
end.

File: day15-1.c
#include "STDIO.H"

hashme(str)
char *str;
{
  int i,current;

  current = 0;
  for (i=0; i<strlen(str); i++) {
    current = current + (int) str[i];
    current = current * 17;
    current = current % 256;
  }

  return(current);
}

main()
{
  FILE *fp;
  int ch, bidx;
  long total;
  char buf[80];

  printf("Advent of Code 2023 Day 15, part 1\n\n");

  total = 0;

  if ((fp = fopen("day15in.txt","r")) == NULL) {
    printf("day15in.txt not found\n");
    exit();
  }

  bidx = 0;
  /* Use agetc rather than getc as this is CP/M */
  while ((ch = agetc(fp)) != EOF) {
    if (ch == ',') {
      buf[bidx] = '\0';
      total = total + (long) hashme(buf);
      bidx = 0;
    }
    else
      buf[bidx++] = ch;
  }

  /* Deal with the last value to hash when EOF is reached */
  if (ch == EOF) {
    /* zap the newline and terminate the string */
    buf[bidx-1] = '\0';
    total = total + (long) hashme(buf);
  }

  printf("Sum of hash values is %ld\n",total);
}
